river/tree/nodes/sgt_nodes.py

```diff
@@ -0,0 +1,223 @@
+import collections
+import math
+import numbers
+import sys
+from typing import Dict, Hashable, Optional, Union
+
+from river import stats
+from river.base.typing import FeatureName
+
+from ..base import Leaf
+from ..splitter.sgt_quantizer import DynamicQuantizer, StaticQuantizer
+from ..utils import BranchFactory, GradHess, GradHessMerit, GradHessStats
+from .branch import NominalMultiwayBranch, NumericBinaryBranch
+
+
+class SGTLeaf(Leaf):
+    """Leaf Node of the Stochastic Gradient Trees (SGT).
+
+    There is only one type of leaf in SGTs. It handles only gradient and hessian
+    information about the target. The tree handles target transformation and encoding.
+
+    Parameters
+    ----------
+    prediction
+        The initial prediction of the leaf.
+    depth
+        The depth of the leaf in the tree.
+    split_params
+        Parameters passed to the feature quantizers.
+    """
+
+    def __init__(self, prediction=0.0, depth=0, split_params=None):
+        super().__init__()
+        self._prediction = prediction
+        self.depth = depth
+
+        # Quantizer params are dynamically updated
+        self.split_params = (
+            split_params if split_params is not None else collections.defaultdict(dict)
+        )
+        self.last_split_attempt_at = 0
+
+        self._split_stats: Optional[
+            Dict[
+                FeatureName,
+                Union[Dict[Hashable, GradHessStats], DynamicQuantizer, StaticQuantizer],
+            ]
+        ] = {}
+        self._update_stats = GradHessStats()
+
+    def reset(self):
+        self._split_stats = {}
+        self._update_stats = GradHessStats()
+
+    def update(self, x: dict, gh: GradHess, sgt, w: float = 1.0):
+        for idx, x_val in x.items():
+            if not isinstance(x_val, numbers.Number) or idx in sgt.nominal_attributes:
```

Comment:
An `is_categorical` method might ease readability.

Reply:
Totally agreed. I added this [method](https://github.com/online-ml/river/blob/b33ed26aa625562c6df2c6e1bda4ebb59616a72c/river/tree/nodes/sgt_nodes.py#L56-L58). I wish it had fewer parameters XD